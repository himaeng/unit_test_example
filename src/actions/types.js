export const LOAD_REPOS_FAILURE = 'load_repos_failure';
export const LOAD_REPOS_REQUEST = 'load_repos_request';
export const LOAD_REPOS_SUCCESS = 'load_repos_success';
export const SELECT_REPO = 'select_repo';
