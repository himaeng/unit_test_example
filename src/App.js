import React, { Component } from 'react';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';

import RepoPage from './pages/RepoPage';
import reducers from './reducers';
import loggerMiddleware from './middlewares/loggerMiddleware';
import asyncMiddleware from './middlewares/asyncMiddleware';

const store = createStore(reducers, {},
  applyMiddleware(
    thunk,
    loggerMiddleware,
    asyncMiddleware
  ));

class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <RepoPage />
      </Provider>
    );
  }
}

export default App
