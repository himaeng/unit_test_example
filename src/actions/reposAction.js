import {
  LOAD_REPOS_REQUEST,
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS_FAILURE,
  SELECT_REPO
} from './types';

export const loadReposSuccess = (data) => {
  return ({
    type: LOAD_REPOS_SUCCESS,
    data
  });
};

export const loadReposRequest = () => {
  return ({
    type: LOAD_REPOS_REQUEST
  });
};

export const loadReposFailure = (error) => {
  return ({
    type: LOAD_REPOS_FAILURE,
    error
  });
};

export const fetchData = (endPoint) => {
  return async (dispatch) => {
    try {
      dispatch(loadReposRequest());
      const response      = await fetch(endPoint);
      const responseJson  = await response.json();
      dispatch(loadReposSuccess(responseJson.items));
    } catch (error) {
      dispatch(loadReposFailure(error));
    }
  };
}

export const selectRepo = (id) => {
  return ({
    type: SELECT_REPO,
    id
  })
}
