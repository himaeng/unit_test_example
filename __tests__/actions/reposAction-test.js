import mockStore from '../../mocks/redux-mock-store';
import { loadReposRequest, selectRepo } from '../../src/actions';

const store = mockStore();

describe('reposAction', () => {
  beforeEach (() => {
    store.clearActions();
  });

  it('should handle reposAction for success case', async () => {
    const response = '{"items": [{"id": 1}]}';
    fetch.mockResponseSuccess(response);

    await store.dispatch(loadReposRequest('/test'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should handle reposAction for failure case', async () => {
    const mockError = new Error('ดูสิแม่ ผมชื่อเออเร่อ');
    fetch.mockResponseFailure(mockError);

    await store.dispatch(loadReposRequest('/test'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should handle selectRepo action', () => {
    store.dispatch(selectRepo(1100));
    expect(store.getActions()).toMatchSnapshot();
  });
});
