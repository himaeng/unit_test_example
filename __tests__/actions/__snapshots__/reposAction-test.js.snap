// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reposAction should handle reposAction for failure case 1`] = `
Array [
  Object {
    "payload": Promise {
      "_45": 0,
      "_54": null,
      "_65": [Error: ดูสิแม่ ผมชื่อเออเร่อ],
      "_81": 2,
    },
    "type": "load_repos_request",
  },
]
`;

exports[`reposAction should handle reposAction for success case 1`] = `
Array [
  Object {
    "payload": Promise {
      "_45": 0,
      "_54": null,
      "_65": Object {
        "json": [Function],
      },
      "_81": 1,
    },
    "type": "load_repos_request",
  },
]
`;

exports[`reposAction should handle selectRepo action 1`] = `
Array [
  Object {
    "id": 1100,
    "type": "select_repo",
  },
]
`;
